launcher/src/main/java/org/apache/spark/launcher/LauncherConnection.java

```diff
@@ -107,4 +106,17 @@ public void close() throws IOException {
     }
   }
 
+  private void close(Exception exp) throws IOException {
+	if (!closed) {
+	  synchronized (this) {
+	    if (!closed) {
+	      handle(new SetState(State.FINISHED_UNKNOWN));
```

Comment:
Hmm, this looks out of place. I think a better place for this is the `ServerConnection` class (in `LauncherServer.java`). That's the class that understands this message. Then your patch is also much simpler since you don't need this second close method, it turns into a single line in `ServerConnection.close` (and maybe a log message).